sdks/python/apache_beam/transforms/trigger_test.py

```diff
@@ -518,6 +519,28 @@ def format_result(k_v):
                   'B-3': {10, 15, 16},
               }.items())))
 
+  def test_never(self):
+    with TestPipeline() as p:
+
+      def construct_timestamped(k_t):
+        return TimestampedValue((k_t[0], k_t[1]), k_t[1])
+
+      def format_result(k_v):
+        return ('%s-%s' % (k_v[0], len(k_v[1])), set(k_v[1]))
+
+      result = (
+          p
+          | beam.Create([1, 1, 2, 3, 4, 5, 10, 11])
+          | beam.FlatMap(lambda t: [('A', t), ('B', t + 5)])
+          | beam.Map(construct_timestamped)
+          | beam.WindowInto(
+              FixedWindows(10),
+              trigger=Never(),
+              accumulation_mode=AccumulationMode.DISCARDING)
+          | beam.GroupByKey()
+          | beam.Map(format_result))
+      assert_that(result, equal_to([]))
```

Comment:
Never trigger is misnamed. It means that the trigger itself never fires, but that at window GC output is produced. That's why PAssert uses it to gather the full result.

Reply:
Ack. (Interestingly, I originally was expecting it to fire at EoW.) The ULR doesn't yet support gc timers. I can look into fixing this (or at least making the "Never" trigger correct). 

As an aside, the non-trivial triggering and windowing in PAssert makes it less than ideal for validating new runners.

Reply:
I couldn't find an existing bug about closing behaviors, so I filed BEAM-10149. That seems pretty invasive to change as part of this PR, so I added a hack to support its use with proper closing behavior in batch only. PTAL.

Reply:
Good point about that in PAssert. Though isn't the triggering just "fire at GC" and the windowing nothing special? Previously, it depended on side inputs which tend to come later in a runner's lifecycle.

Reply:
I think even the concept of GC triggers (which have to be scheduled at a different time than a "normal" trigger at EOW) and late data is rather advanced. Also, yes, it used a custom (java serializable) WindowFn (didn't investigate which one this was). Possibly the right thing to do here would be to use "standard" triggers for the normal case, and advanced features when needed, assuming they can be queried at pipeline construction time. But this is completely an aside. 

Thanks for the review. I discovered another couple of issues when running the full suite; I'll get them into separate PRs.